greenplumpython/experimental/file.py

```diff
@@ -0,0 +1,79 @@
+import base64
+import inspect
+import io
+import pathlib
+import tarfile
+import uuid
+from typing import get_type_hints
+
+import psycopg2
+
+import greenplumpython as gp
+from greenplumpython.func import NormalFunction
+
+_CHUNK_SIZE = 256 * 1024 * 1024  # Must be much < 1 GB
+
+
+@gp.create_function
+def _dump_file_chunk(tmp_archive_name: str, chunk_base64: str) -> int:
+    tmp_archive_base = pathlib.Path("/") / "tmp" / tmp_archive_name
+    tmp_archive_path = tmp_archive_base / f"{tmp_archive_name}.tar.gz"
+    with open(tmp_archive_path, "ab") as tmp_archive:
+        tmp_archive.write(base64.b64decode(chunk_base64))
+    return 0
+
+
+@gp.create_function
+def _extract_files(tmp_archive_name: str) -> list[str]:
+    tmp_archive_base = pathlib.Path("/") / "tmp" / tmp_archive_name
+    tmp_archive_path = tmp_archive_base / f"{tmp_archive_name}.tar.gz"
+    extracted_dir = tmp_archive_base / "extracted"
+    if not extracted_dir.exists():
+        with tarfile.open(tmp_archive_path, "r:gz") as tmp_archive:
+            extracted_dir.mkdir()
+            tmp_archive.extractall(str(extracted_dir))
+        tmp_archive_path.unlink()
+    for path in extracted_dir.rglob("*"):
+        if path.is_file() and not path.is_symlink():
+            yield str(path.resolve())
+
+
+@classmethod
+def _from_files(_, files: list[str], parser: NormalFunction, db: gp.Database) -> gp.DataFrame:
+    tmp_archive_name = f"tar_{uuid.uuid4().hex}"
+    tmp_archive_base = pathlib.Path("/") / "tmp" / tmp_archive_name
+    tmp_archive_base.mkdir()
+    tmp_archive_path = tmp_archive_base / f"{tmp_archive_name}.tar.gz"
+    with tarfile.open(tmp_archive_path, "w:gz") as tmp_archive:
+        for file_path in files:
+            tmp_archive.add(pathlib.Path(file_path))
+    with psycopg2.connect(db._conn.dsn, options="-c gp_session_role=utility") as util_conn:
```

Comment:
The `GreenplumPython` needs to support pg as well. `gp_session_role` won't be needed for pg. I am not sure if an error will be reported if pg doesn't recognize the option,